pkg/tfbridge/property_path.go

```diff
@@ -230,22 +229,66 @@ func propertyValueTriggersReplacement(
 	return replacement
 }
 
-// pathContainsComputed returns true if the schema contains a Computed property at a path prefixed by path.
-func pathContainsComputed(
-	path propertyPath, rootTFSchema shim.SchemaMap, rootPulumiSchema map[string]*info.Schema,
+// validInputsFromPlan returns true if the given plan property value could originate from the given inputs.
+// Under the hood, it walks the plan and the inputs and checks that all differences stem from computed properties.
+// Any differences coming from properties which are not computed will be rejected.
+// Note that we are relying on the fact that the inputs will have defaults already applied.
+func validInputsFromPlan(
+	path propertyPath,
+	inputs resource.PropertyValue,
+	plan resource.PropertyValue,
+	tfs shim.SchemaMap,
+	ps map[string]*info.Schema,
 ) bool {
-	tfs, _, err := lookupSchemas(path, rootTFSchema, rootPulumiSchema)
-	if err != nil {
-		return false
-	}
+	abortErr := errors.New("abort")
+	visitor := func(
+		subpath propertyPath, inputsSubVal, planSubVal resource.PropertyValue,
+	) error {
+		// Do not compare and do not descend into internal properties.
+		if subpath.IsReservedKey() {
+			return SkipChildrenError{}
+		}
 
-	computed := false
-	visitor := func(path walk.SchemaPath, tfs shim.Schema) {
-		if tfs.Computed() {
-			computed = true
+		tfs, _, err := lookupSchemas(subpath, tfs, ps)
+		if err != nil {
+			// TODO log
+			return abortErr
+		}
+
+		if tfs.Computed() && inputsSubVal.IsNull() {
+			// This is a computed property populated by the plan. We should not recurse into it.
+			return SkipChildrenError{}
+		}
+
+		if tfs.Type() == shim.TypeList || tfs.Type() == shim.TypeMap {
+			// We only need to check the leaf values, so we skip any collection types.
+			return nil
 		}
-	}
-	walk.VisitSchema(tfs, visitor)
 
-	return computed
+		// We can not descend into nested sets as planning re-orders the elements
```

Comment:
This seems very approximate. So if we happen to match via DeepEquals we presume accept (validInputsFromPlan==>true), but if reordering happened and we did not match by DeepEquals we fail (validInputsFromPlan==>false)? This may be in some sense acceptable but good to call out in the doc comment.

Reply:
It seems good enough for now. Nested sets are rare and we do not fail completely here - we will just show a slightly more verbose diff from the engine.